runtime-modules/proposals/engine/src/types/mod.rs

```diff
@@ -279,8 +280,43 @@ where
         actual_votes_fraction.deconstruct() >= approval_quorum_fraction.deconstruct()
     }
 
+    // Verifies that approval threshold is still achievable for the proposal.
+    // Compares actual approval votes sum with remaining possible votes number.
+    fn is_approval_threshold_achievable(&self) -> bool {
+        let remaining_votes_count = self.total_voters_count - self.votes_count;
+        let possible_approval_votes_fraction = Perbill::from_rational_approximation(
+            self.approvals + remaining_votes_count,
+            self.votes_count + remaining_votes_count,
+        );
+
+        let required_threshold_fraction =
+            Perbill::from_percent(self.proposal.parameters.approval_threshold_percentage);
+
+        possible_approval_votes_fraction.deconstruct() >= required_threshold_fraction.deconstruct()
+    }
+
+    // Verifies that both approval and slashing thresholds cannot be achieved.
+    pub fn is_rejection_imminent(&self) -> bool {
```

Comment:
I think this is too strict, remember, rejection is imminent if one cannot satisfy the constraint on both quorum AND threshold. This is equivalent to failing on one or the other (or both), for both approval and slashing. You are now requiring that you specifically cannot achieve the threshold, but this can in principle not be determinable as early.

Reply:
Quorum calculation could be omitted because when dealing with for example "approve threshold achievability" we calculate possible outcome when all remaining councilors vote "Approve". Obviously 100% of votes satisfy any quorum requirements.

Rejection is imminent when (approve threshold is not achievable) AND (slash threshold is not achievable).

For example (test proposal_status_resolution_approval_achievable_works_correctly):
    Given:

```
    slashing_threshold_percentage: 50
    approval_threshold_percentage: 50
    council size: 600
```

    1. Only one councilor approves.
    2. 301 other councilors abstain.

So, even if the remaining 298 councilors also approve the proposal the approval threshold can not be achieved.

Thus, having 1 approval and 300 abstains we could decide that the proposal should be rejected.

Reply:
I don't fully understand your reply, but is your point that its actually never possible to rule out early that quorum is unsatisfiable, because everyone could still show up in time to satisfy any quorum requirement. The only thing you can rule out early is that a given threshold could be satisfied?

Reply:
Correct.